cpp/src/arrow/util/io_util.cc

```diff
@@ -1867,7 +1867,16 @@ Result<std::unique_ptr<TemporaryDir>> TemporaryDir::Make(const std::string& pref
       [&](const NativePathString& base_dir) -> Result<std::unique_ptr<TemporaryDir>> {
     Status st;
     for (int attempt = 0; attempt < 3; ++attempt) {
-      PlatformFilename fn(base_dir + kNativeSep + base_name + kNativeSep);
+      // Note: certain temporary directories of MacOS contains a trailing slash
+      // Handling the base_dir with trailing slash
+      PlatformFilename fn;
+      if (base_dir.back() == kNativeSep) {
+        PlatformFilename fn_base_dir(base_dir);
+        PlatformFilename fn_base_name(base_name + kNativeSep);
+        fn = fn_base_dir.Join(fn_base_name);
+      } else {
+        fn = PlatformFilename(base_dir + kNativeSep + base_name + kNativeSep);
+      }
```

Comment:
```suggestion
      PlatformFilename fn_base_dir(base_dir);
      fn = fn_base_dir.Join(base_name) + kNativeSep;
```

Does this work?  `Join` already has logic in place to test for a trailing separator so I'm not sure why we need to also test for one here.